src/frequenz/dispatch/actor.py

```diff
@@ -0,0 +1,199 @@
+                    self._update_dispatch_schedule(dispatch, old_dispatch)
+                    logging.info("Updated dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(dispatch)
+
+        except grpc.aio.AioRpcError as error:
+            logging.error("Error fetching dispatches: %s", error)
+            self._dispatches = old_dispatches
+            return
+
+        for dispatch in old_dispatches.values():
+            logging.info("Deleted dispatch: %s", dispatch)
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+    def _update_dispatch_schedule(
+        self, dispatch: Dispatch, old_dispatch: Dispatch | None
+    ) -> None:
+        """Update the schedule for a dispatch.
+
+        Schedules, reschedules or cancels the dispatch based on the start_time
+        and active status.
+
+        Args:
+            dispatch: The dispatch to update the schedule for.
+            old_dispatch: The old dispatch, if available.
+        """
+        if (
+            old_dispatch
+            and old_dispatch.active
+            and old_dispatch.start_time != dispatch.start_time
+        ):
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+        if dispatch.active and dispatch.id not in self._scheduled:
+            self._scheduled[dispatch.id] = asyncio.create_task(
+                self._schedule_task(dispatch)
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
+        """Wait for a dispatch to become ready.
+
+        Waits for the dispatches next run and then notifies that it is ready.
+
+        Args:
+            dispatch: The dispatch to schedule.
+        """
+        while True:
+            next_time = self.calculate_next_run(
+                dispatch, datetime.now().replace(tzinfo=timezone.utc)
+            )
+            if next_time is None:
+                logging.info("Dispatch finished: %s", dispatch)
+                self._scheduled.pop(dispatch.id)
+                return
+
+            if (
+                next_time - datetime.now().replace(tzinfo=timezone.utc)
+                > _MAX_AHEAD_SCHEDULE
+            ):
+                await asyncio.sleep(_MAX_AHEAD_SCHEDULE.total_seconds())
+                continue
+
+            logging.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
+            logging.info(
+                "Sleeping for %s",
+                next_time - datetime.now().replace(tzinfo=timezone.utc),
+            )
+            await asyncio.sleep(
+                (
+                    next_time - datetime.now().replace(tzinfo=timezone.utc)
+                ).total_seconds()
+            )
+            logging.info("Dispatch ready: %s", dispatch)
+            await self._ready_dispatch_sender.send(dispatch)
+
+    @classmethod
+    def calculate_next_run(cls, dispatch: Dispatch, _from: datetime) -> datetime | None:
```

Comment:
why does this have to be a public class method?  from what I can see, it will be used only from within the `Dispatcher`, and only from contexts where a `self` is available.

Reply:
It doesn't have to be, but as a user who wants to create new dispatches and investigate existing ones, I figured it's very useful to know when the next time is a dispatch would run.